sdk/identity/azure-identity/azure/identity/_internal/msal_credentials.py

```diff
@@ -57,21 +89,53 @@ def _get_app(self):
         # type: () -> msal.ClientApplication
         pass
 
-    def _create_app(self, cls):
-        # type: (Type[msal.ClientApplication]) -> msal.ClientApplication
+    def _create_app(self, cls, **kwargs):
+        # type: (Type[msal.ClientApplication], **Any) -> msal.ClientApplication
         """Creates an MSAL application, patching msal.authority to use an azure-core pipeline during tenant discovery"""
 
         # MSAL application initializers use msal.authority to send AAD tenant discovery requests
         with self._adapter:
             # MSAL's "authority" is a URL e.g. https://login.microsoftonline.com/common
-            app = cls(client_id=self._client_id, client_credential=self._client_credential, authority=self._base_url)
+            app = cls(
+                client_id=self._client_id,
+                client_credential=self._client_credential,
+                authority=self._base_url,
+                token_cache=self._cache,
+                **kwargs
+            )
 
         # monkeypatch the app to replace requests.Session with MsalTransportAdapter
         app.client.session.close()
         app.client.session = self._adapter
 
         return app
 
+    def _account_matches_profile(self, account):
+        return (
+            account.get("home_account_id") == self._profile.home_account_id
+            and account.get("username") == self._profile.username
+            and account.get("environment") == self._profile.environment
+            and account.get("realm") == self._profile.tenant_id
```

Comment:
> and account.get("realm") == self._profile.tenant_id [](start = 11, length = 52)

this check is incorrect based on my test, the value of account.get('realm') is 'organizations'. I think we do not need this check since home_account_id include the tenant id

Reply:
if I specify tenant_id like
credential, profile = InteractiveBrowserCredential.authenticate(
           # silent_auth_only=True
            scope = '[https://management.azure.com/.default](https://management.azure.com/.default)',
            tenant_id = tenant
        )
The value of realm is the tenant_id. But in this scenario, the cached account still not match the profile. The home_account_id in the profile is correct, but the home_account_id in the MSAL cache is my organization tenants's home_account_id

---
In reply to: [396249087](https://github.com/chlowell/azure-sdk-for-python/pull/4#discussion_r396249087) [](ancestors = 396249087)

Reply:
You're correct we needn't check the realm, thank you for pointing that out.